Mathlib/Combinatorics/Schnirelmann.lean

```diff
@@ -0,0 +1,247 @@
+  rw [le_iff_forall_pos_lt_add]
+  simp only [schnirelmannDensity_lt_iff]
+
+/-- The Schnirelmann density is unaffected by adding `0`. -/
+lemma schnirelmannDensity_insert_zero [DecidablePred (· ∈ insert 0 A)] :
+    schnirelmannDensity (insert 0 A) = schnirelmannDensity A :=
+  by rw [schnirelmannDensity, schnirelmannDensity]; congr; ext ⟨n, hn⟩; congr 3; ext x; aesop
+
+/-- The Schnirelmann density is unaffected by removing `0`. -/
+lemma schnirelmannDensity_diff_singleton_zero [DecidablePred (· ∈ A \ {0})] :
+    schnirelmannDensity (A \ {0}) = schnirelmannDensity A :=
+  by rw [schnirelmannDensity, schnirelmannDensity]; congr; ext ⟨n, hn⟩; congr 3; ext x; aesop
+
+/--
+If the Schnirelmann density is `0`, there is a positive natural for which
+`|A ∩ {1, ..., n}| / n < ε`, for any positive `ε`.
+Note this cannot be improved to `∃ᶠ n : ℕ in at_top`, as can be seen by `A = {1}`.
+-/
+lemma exists_of_schnirelmannDensity_eq_zero {ε : ℝ} (hε : 0 < ε) (hA : schnirelmannDensity A = 0) :
+    ∃ n, 0 < n ∧ ((Ioc 0 n).filter (· ∈ A)).card / n < ε := by
+  by_contra' h
+  rw [←le_schnirelmannDensity_iff] at h
+  linarith
+
+end
+
+@[simp] lemma schnirelmannDensity_empty : schnirelmannDensity ∅ = 0 :=
+  schnirelmannDensity_eq_zero_of_one_not_mem (by simp)
+
+/-- The Schnirelmann density of any finset is `0`. -/
+lemma schnirelmannDensity_finset (A : Finset ℕ) : schnirelmannDensity A = 0 := by
+  refine le_antisymm ?_ schnirelmannDensity_nonneg
+  simp only [schnirelmannDensity_le_iff_forall, zero_add]
+  intro ε hε
+  wlog hε₁ : ε ≤ 1 generalizing ε
+  · obtain ⟨n, hn, hn'⟩ := this 1 zero_lt_one le_rfl
+    exact ⟨n, hn, hn'.trans_le (le_of_not_le hε₁)⟩
+  let n : ℕ := ⌊A.card / ε⌋₊ + 1
+  have hn : 0 < n := Nat.succ_pos _
+  use n, hn
+  rw [div_lt_iff (Nat.cast_pos.2 hn), ←div_lt_iff' hε, Nat.cast_add_one]
+  exact (Nat.lt_floor_add_one _).trans_le' <| div_le_div_of_le hε.le <| Nat.cast_le.2 <|
+    card_le_of_subset <| by simp [subset_iff]
+
+/-- The Schnirelmann density of any finite set is `0`. -/
+lemma schnirelmannDensity_finite {A : Set ℕ} [DecidablePred (· ∈ A)] (hA : A.Finite) :
+    schnirelmannDensity A = 0 := by simpa using schnirelmannDensity_finset hA.toFinset
+
+@[simp] lemma schnirelmannDensity_univ : schnirelmannDensity Set.univ = 1 :=
+  (schnirelmannDensity_eq_one_iff_of_zero_mem (by simp)).2 (by simp)
+
+lemma schnirelmannDensity_setOf_even : schnirelmannDensity (setOf Even) = 0 :=
+  schnirelmannDensity_eq_zero_of_one_not_mem (by simp)
+
+lemma schnirelmannDensity_setOf_prime : schnirelmannDensity (setOf Nat.Prime) = 0 :=
+  schnirelmannDensity_eq_zero_of_one_not_mem (by simp)
+
+/--
+The Schnirelmann density of the set of naturals which are `1 mod m` is `1 / m`, for any `m ≠ 1`.
+
+Note that if `m = 1`, this set is empty.
+-/
+lemma schnirelmannDensity_setOf_mod_eq_one {m : ℕ} (hm : m ≠ 1) :
+    schnirelmannDensity {n | n % m = 1} = 1 / m := by
```

Comment:
Maybe better written using `Nat.ModEq`? Then it's also true for `m = 1`.

Reply:
I added the ModEq version as well, but kept this version since I think both can be useful.